### ISSUE TITLE
Build errors

### ISSUE THREAD

**Issue by CiaranWelsh** [sys-bio/libroadrunner-deps#12]
I deleted my existing libroadrunner-deps repository to start from fresh. The first time the cmake configure command is run on Windows it cannot find zlibstatic. This is the warning message: 
```
CMake Error at D:/roadrunner/libroadrunner-deps/cmake-build-release/CMakeFiles/CMakeTmp/CMakeLists.txt:17 (target_link_libraries):
  Error evaluating generator expression:

    $<TARGET_LINKER_FILE:zlibstatic>

  No target "zlibstatic"


CMake Error at C:/Program Files/CMake/share/cmake-3.18/Modules/CheckLibraryExists.cmake:67 (try_compile):
  Failed to generate test project build system.
```
We *may be able to solve this with a call to `add_dependency` somewhere. However, when you run the configure command again, zlibstatic is found but we run into other problems. Specifically, sbml cannot find the extension packages: 

```
CMake Error at third_party/libsbml/src/comp-package.cmake:36 (include):
  include could not find load file:

    D:/roadrunner/libroadrunner-deps/comp-package.cmake
Call Stack (most recent call first):
  third_party/libsbml/src/CMakeLists.txt:51 (include)


CMake Error at third_party/libsbml/src/distrib-package.cmake:40 (include):
  include could not find load file:

    D:/roadrunner/libroadrunner-deps/distrib-package.cmake
Call Stack (most recent call first):
  third_party/libsbml/src/CMakeLists.txt:51 (include)

...
```

Thirdly, I get the following warning in CMake: 

```
CMake Warning (dev) at third_party/libsbml/CMakeLists.txt:35 (project):
  Policy CMP0048 is not set: project() command manages VERSION variables.
  Run "cmake --help-policy CMP0048" for policy details.  Use the cmake_policy
  command to set the policy and suppress this warning.

  The following variable(s) would be set to empty:

    PROJECT_VERSION
    PROJECT_VERSION_MAJOR
    PROJECT_VERSION_MINOR
    PROJECT_VERSION_PATCH
```

This may be resolvable by updating the cmake minimum version within libsbml. Is this a possibility?

**Comment by luciansmith** [comment]
I'm not currently worried about the warnings, and I agree that I think the zlibstatic thing can be solved with dependencies.  But looking for the package cmake files in the wrong directory is truly strange to me.  Everything works on Windows for me, though I'm probably using a more recent version of CMake?  I'll investigate.

**Comment by CiaranWelsh** [comment]
Ah, maybe libsbml is using ${CMAKE_SOURCE_DIR} somewhere, which when libsbml is embedded in another project, does not point to the top level cmakelists of libsbml. This is why I never use CMAKE_SOURCE_DIR and instead prefer CMAKE_CURRENT_SOURCE_DIR. Though if this were the problem then I don't understand why this wasn't a bug in your system?

**Comment by luciansmith** [comment]
So, the include line in question is in libsbml/src/comp-package.cmake :

include(${LIBSBML_ROOT_SOURCE_DIR}/comp-package.cmake)

and is intended to point to libsbml/comp-package.cmake

In turn, LIBSBML_ROOT_SOURCE_DIR is set in libsbml/CMakeLists.txt:

set(LIBSBML_ROOT_SOURCE_DIR "${CMAKE_CURRENT_SOURCE_DIR}" CACHE PATH
    "Path to the libSBML root source directory")

This works for me without issue, so I'm confused as to why you'd have problems!

We could set LIBSBML_ROOT_SOURCE_DIR ourselves, perhaps?  I'm not sure how that would make a difference, though.

**Comment by CiaranWelsh** [comment]
Hmm I didn’t dig into this deeply as I was focusing on figuring out how to do what we want in azure. However, it’s not just my computer because the azure for the deps package does this as well https://dev.azure.com/TheRoadrunnerProject/libroadrunner-deps/_build/results?buildId=246&view=results. Could it be worth building the deps in a clean environment? Docker or vm to reproduce the problem?

**Comment by luciansmith** [comment]
What is the cmake command that you are sending to azure or typing at the command line?  I'm running cmake-gui with no arguments, and hitting 'compile' twice.

**Comment by luciansmith** [comment]
Wait, I got it!  The CMakeLists.txt file I'm using is different from the official release.  I thought I checked it in, but I guess not?  Anyway, fixing...

(The LIBSBML_ROOT_SOURCE_DIR is *not* being set in libsbml to the current source dir, but the root.)

**Comment by CiaranWelsh** [comment]
Aaah there's the bugger. I wont paste my configure arguments then - they were the same as yours anyway. When you push to the master branch the "TheRoadrunnerProject/libroadrunner-deps" pipeline should go ahead and build. (You can find this anytime by clicking on the "azure-pipelines" badge from the libroadrunner-deps github page, if you didn't already know this). 

Not strictly on topic but I'll add some code which will publish the resulting builds into easily downloadable locations. My plan is to figure out how to use these published releases directly, rather than needing to manually publish a release then download the zip.

**Comment by luciansmith** [comment]
OK, this is dumb, but I've updated my copy of libsbml to point at the 'fix-cmake-root-dirs' branch (and had done so several weeks ago), but somehow I can't get git to commit it properly.  Would you be able to do that?

**Comment by CiaranWelsh** [comment]
Sure, for reference, this is how I did it: 

```
(base) PS D:\roadrunner\libroadrunner-deps> git branch --all
* master
  remotes/origin/32-bit
  remotes/origin/HEAD -> origin/master
  remotes/origin/ciaran-libsbml-as-submodule
  remotes/origin/develop
  remotes/origin/libsbml-5.16.0
  remotes/origin/libsbml-5.17.0
  remotes/origin/master
  remotes/origin/vs14-x64
(base) PS D:\roadrunner\libroadrunner-deps> cd .\third_party\libsbml\
(base) PS D:\roadrunner\libroadrunner-deps\third_party\libsbml> git status
HEAD detached at 010b92344
nothing to commit, working tree clean
(base) PS D:\roadrunner\libroadrunner-deps\third_party\libsbml> git branch --all
* (HEAD detached at 010b92344)
  development
  remotes/origin/HEAD -> origin/development
  remotes/origin/development
  remotes/origin/fix-cmake-root-dirs
  remotes/origin/fix-sonarcloud-smells-2
  remotes/origin/fixing-sonarcloud-smells
  remotes/origin/stable
(base) PS D:\roadrunner\libroadrunner-deps\third_party\libsbml> git checkout fix-cmake-root-dirs
Previous HEAD position was 010b92344 Merge pull request #106 from sbmlteam/update-readme-for-5.19
Switched to a new branch 'fix-cmake-root-dirs'
Branch 'fix-cmake-root-dirs' set up to track remote branch 'fix-cmake-root-dirs' from 'origin'.
(base) PS D:\roadrunner\libroadrunner-deps\third_party\libsbml> git status
On branch fix-cmake-root-dirs
Your branch is up to date with 'origin/fix-cmake-root-dirs'.

nothing to commit, working tree clean
(base) PS D:\roadrunner\libroadrunner-deps\third_party\libsbml> cd ..
(base) PS D:\roadrunner\libroadrunner-deps\third_party> git status
On branch master
Your branch is ahead of 'origin/master' by 1 commit.
  (use "git push" to publish your local commits)

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
        modified:   libsbml (new commits)

no changes added to commit (use "git add" and/or "git commit -a")
(base) PS D:\roadrunner\libroadrunner-deps\third_party> git add .\libsbml\
(base) PS D:\roadrunner\libroadrunner-deps\third_party> git commit -m "Point libsbml dependency at fix-cmake-root-dirs branch"
[master 869a190b] Point libsbml dependency at fix-cmake-root-dirs branch
 1 file changed, 1 insertion(+), 1 deletion(-)
(base) PS D:\roadrunner\libroadrunner-deps\third_party> git push
Enumerating objects: 10, done.
Counting objects: 100% (10/10), done.
Delta compression using up to 12 threads
Compressing objects: 100% (7/7), done.
Writing objects: 100% (7/7), 963 bytes | 963.00 KiB/s, done.
Total 7 (delta 5), reused 0 (delta 0), pack-reused 0
remote: Resolving deltas: 100% (5/5), completed with 3 local objects.
To https://github.com/sys-bio/libroadrunner-deps.git
   714c653f..869a190b  master -> master
(base) PS D:\roadrunner\libroadrunner-deps\third_party> git status
On branch master
Your branch is up to date with 'origin/master'.

nothing to commit, working tree clean
(base) PS D:\roadrunner\libroadrunner-deps\third_party> cd .\libsbml\
(base) PS D:\roadrunner\libroadrunner-deps\third_party\libsbml> git branch
  development
* fix-cmake-root-dirs
(base) PS D:\roadrunner\libroadrunner-deps\third_party\libsbml>
```

**Comment by CiaranWelsh** [comment]
... building https://dev.azure.com/TheRoadrunnerProject/libroadrunner-deps/_build?definitionId=9&_a=summary

**Comment by CiaranWelsh** [comment]
So clearly that failed... 

I'm looking at the cmake source code and it appears the file at 

    D:\roadrunner\libroadrunner-deps\third_party\libsbml\comp-package.cmake

tries to include itself on line 36.

**Comment by luciansmith** [comment]
No, it's including the file of the same name one level up.  And actually, that part seems to be fine; configure now passes, and it's the build that then fails.

**Comment by CiaranWelsh** [comment]
Okay I got this package compiling locally now, here is the link for the remote build: https://dev.azure.com/TheRoadrunnerProject/libroadrunner-deps/_build/results?buildId=248&view=results

**Comment by luciansmith** [comment]
OK, it's now failing on Linux due to:
```
2020-12-11T21:41:40.2267811Z /home/vsts/work/1/s/third_party/libsbml/src/sbml/common/libsbml-version.cpp:56:10: fatal error: libxml/xmlversion.h: No such file or directory
2020-12-11T21:41:40.2269324Z  #include <libxml/xmlversion.h>
2020-12-11T21:41:40.2270109Z           ^~~~~~~~~~~~~~~~~~~~~
2020-12-11T21:41:40.2272181Z compilation terminated.
```

This is, I believe, a file that is generated by CMake and then installed.  I fixed this for me by adding this to third-party/CMakeLists.txt:

```
set(LIBXML_INCLUDE_DIR "${CMAKE_CURRENT_SOURCE_DIR}/libxml2/include")
set(EXTRA_INCLUDE_DIRS "${EXTRA_INCLUDE_DIRS}" "${CMAKE_CURRENT_BINARY_DIR}/libxml2/include/" "${CMAKE_CURRENT_BINARY_DIR}/third_party/libxml2/include/")
```

However, perhaps the different Azure environment means we need to solve it a different way there?

**Comment by CiaranWelsh** [comment]
I've just pushed a fix for this problem. The issue is 
    
    set(LIBXML_LIBRARY $<TARGET_LINKER_FILE:LibXml2>)

which evaluates to the location of the libxml2 library at build time should instead be

    set(LIBXML_LIBRARY LibXml2)

LibXml2, the target we are using for libxml2, has its include directories set to PUBLIC, so anything that links to this target will also have the necessary INCLUDE_DIRECTORIES. Hopefully this will solve the problem.

**Comment by luciansmith** [comment]
Sweet, looks like that fixes the linux and mac builds!

The Windows build is still complaining instantly when it tries to build:

```
"C:\windows\system32\cmd.exe" /D /E:ON /V:OFF /S /C "CALL "D:\a\_temp\a1fdfa76-526c-414a-bbdf-5d5b1658c79c.cmd""
Microsoft (R) Build Engine version 16.8.2+25e4d540b for .NET Framework
Copyright (C) Microsoft Corporation. All rights reserved.

C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\MSBuild\Microsoft\VC\v160\Microsoft.CppBuild.targets(408,5): error MSB8013: This project doesn't contain the Configuration and Platform combination of $(BUILD_TYPE)|x64. [D:\a\1\s\build\ZERO_CHECK.vcxproj]
##[error]Cmd.exe exited with code '1'.
Finishing: BuildCommand
```

**Comment by CiaranWelsh** [comment]
Yep - nearly there. I think windows is complaining because of azure problems not actual build problems. Its an issue with the code for publishing the release. Fingers crossed for the next build https://dev.azure.com/TheRoadrunnerProject/libroadrunner-deps/_build/results?buildId=252&view=results

**Comment by luciansmith** [comment]
Yup, you got it!  Thank you again.

**Comment by CiaranWelsh** [comment]
No problem. 

Issue fixed with 834f0fbc1b4aa6034b1efddddfcbadcb17ba8b23  and downloadable releases are [here](https://dev.azure.com/TheRoadrunnerProject/libroadrunner-deps/_build/results?buildId=254&view=artifacts&pathAsName=false&type=publishedArtifacts)
